Panic when there is a query execution related error

If one query fails for whatever reason the application panics and quits.

Sample error output:
```
panic: a previously registered descriptor with the same fully-qualified name as Desc{fqName: "query_result_test", help: "Result of an SQL query", constLabels: {instance="test"}, variableLabels: []} has different label names or a different help string

goroutine 78 [running]:
app/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc420012540, 0xc42012c180, 0x1, 0x1)
        /go/src/app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:404 +0x9e
app/vendor/github.com/prometheus/client_golang/prometheus.MustRegister(0xc42012c180, 0x1, 0x1)
        /go/src/app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:152 +0x53
main.(*QueryResult).RegisterMetrics.func1(0xc420512240, 0x34, 0x91f280, 0xc420012240)
        /go/src/app/set.go:158 +0x13e
main.(*QueryResult).RegisterMetrics(0xc420250330, 0xc4205140f0)
        /go/src/app/set.go:162 +0x27c
main.(*Worker).SetMetrics(0xc42016db90, 0xc420128100, 0x1, 0x4)
        /go/src/app/worker.go:46 +0xf0
main.(*Worker).Fetch(0xc42016db90, 0x7ffd304a5f36, 0x17, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/app/worker.go:117 +0x900
main.(*Worker).Start.func1()
        /go/src/app/worker.go:124 +0x45
main.(*Worker).Start(0xc42016db90, 0x7ffd304a5f36, 0x17)
        /go/src/app/worker.go:131 +0x77
created by main.main
        /go/src/app/main.go:93 +0x70e
```

I'm working on a patch.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -35,8 +35,8 @@ jobs:
 
       - name: Check formatting
         run: |
-            chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
-            $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
+          chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
+          $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
 
       - name: Linting
         if: matrix.os == 'linux'

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -2,7 +2,7 @@ name: CI
 
 on:
   push:
-    branches: '*'
+    branches: [ master ]
   pull_request:
     branches: [ master ]
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -30,13 +30,18 @@ jobs:
       - name: Check out code into the Go module directory
         uses: actions/checkout@v2
 
+      - name: Verify go modules
+        run: go mod verify
+
       - name: Check formatting
         run: |
             chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
             $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
 
-      - name: Verify go modules
-        run: go mod verify
+      - name: Linting
+        run: |
+          go get -u golang.org/x/lint/golint
+          golint .
 
       - name: Get short commit hash and date
         id: get_git_info

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -39,6 +39,7 @@ jobs:
             $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
 
       - name: Linting
+        if: matrix.os == 'linux'
         run: |
           go get -u golang.org/x/lint/golint
           golint .

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -67,6 +67,7 @@ jobs:
             $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
 
       - name: Linting
+        if: matrix.os == 'linux'
         run: |
           go get -u golang.org/x/lint/golint
           golint .

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -58,13 +58,18 @@ jobs:
       - name: Check out code into the Go module directory
         uses: actions/checkout@v2
 
-      - name: Check formatting
-        run: |
-          chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
-          $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
-
       - name: Verify go modules
         run: go mod verify
+
+      - name: Check formatting
+        run: |
+            chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
+            $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
+
+      - name: Linting
+        run: |
+          go get -u golang.org/x/lint/golint
+          golint .
 
       - name: Get short commit hash and date
         id: get_git_info

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -63,8 +63,8 @@ jobs:
 
       - name: Check formatting
         run: |
-            chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
-            $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
+          chmod +x $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
+          $GITHUB_WORKSPACE/scripts/gofmtcheck.sh
 
       - name: Linting
         if: matrix.os == 'linux'

--- a/config_test.go
+++ b/config_test.go
@@ -56,7 +56,7 @@ func Test_loadConfig(t *testing.T) {
 			args: args{file: "test-resources/config-test/datasource-one.yml"},
 			want: &Config{Defaults: createDefaultsData(),
 				DataSources: map[string]DataSource{
-					"mysql-test": DataSource{
+					"mysql-test": {
 						Driver: "mysql",
 						Properties: map[string]interface{}{
 							"host":     "localhost",
@@ -75,7 +75,7 @@ func Test_loadConfig(t *testing.T) {
 			args: args{file: "test-resources/config-test/datasource-two.yml"},
 			want: &Config{Defaults: createDefaultsData(),
 				DataSources: map[string]DataSource{
-					"mysql-test": DataSource{
+					"mysql-test": {
 						Driver: "mysql",
 						Properties: map[string]interface{}{
 							"host":     "localhost",
@@ -85,7 +85,7 @@ func Test_loadConfig(t *testing.T) {
 							"database": "test",
 						},
 					},
-					"mysql-test-2": DataSource{
+					"mysql-test-2": {
 						Driver: "mysql-2",
 						Properties: map[string]interface{}{
 							"host":     "localhost-2",
@@ -139,7 +139,7 @@ func Test_loadQueryConfig(t *testing.T) {
 				config:      c,
 			},
 			want: []*Query{
-				&Query{
+				{
 					Name:          "query_ds_1",
 					DataSourceRef: "my-ds-1",
 					Driver:        "mysql",
@@ -157,7 +157,7 @@ func Test_loadQueryConfig(t *testing.T) {
 					DataField:    "",
 					ValueOnError: "0",
 				},
-				&Query{
+				{
 					Name:          "query_ds_2",
 					DataSourceRef: "my-ds-2",
 					Driver:        "postgresql",
@@ -185,7 +185,7 @@ func Test_loadQueryConfig(t *testing.T) {
 				config:      c,
 			},
 			want: []*Query{
-				&Query{
+				{
 					Name:          "query_ds_1",
 					SQL:           `select 1 as "sum", 2 as "count" from dual`,
 					DataSourceRef: "my-ds-1",
@@ -217,7 +217,7 @@ func Test_loadQueryConfig(t *testing.T) {
 				config:      newConfig(),
 			},
 			want: []*Query{
-				&Query{
+				{
 					Name:   "num_products",
 					Driver: "postgresql",
 					Connection: map[string]interface{}{

--- a/examples/working_example/queries.yml
+++ b/examples/working_example/queries.yml
@@ -23,7 +23,7 @@
 # - response_time_sum with rt as value
 - response_time:
     sql: >
-        select count(*) as cnt, sum(response_time) as rt from requests
+        select count(*) as cnt, sum(response_time) as rt from Requests
     sub-metrics:
       count: cnt
       sum: rt

--- a/examples/working_example/schema.sql
+++ b/examples/working_example/schema.sql
@@ -10,3 +10,14 @@ insert into Companies (name, country) values ('Company1', 'USA');
 insert into Companies (name, country) values ('Company1', 'SWE');
 insert into Companies (name, country) values ('Company1', 'IRL');
 insert into Companies (name, country) values ('Company2', 'IRL');
+
+
+create table Requests (
+	id INTEGER NOT NULL AUTO_INCREMENT,
+	response_time INTEGER NOT NULL,
+	PRIMARY KEY (ID)
+);
+insert into Requests (response_time) values (23);
+insert into Requests (response_time) values (3);
+insert into Requests (response_time) values (7);
+insert into Requests (response_time) values (45);

--- a/main.go
+++ b/main.go
@@ -81,6 +81,7 @@ func main() {
 	// Shared context. Close the cxt.Done channel to stop the workers.
 	ctx, cancel := context.WithCancel(context.Background())
 
+	type contextKey string
 	var w *Worker
 
 	mux := http.NewServeMux()
@@ -89,7 +90,7 @@ func main() {
 		// Create a new worker and start it in its own goroutine.
 		// type key string
 		// const wgKey key = "wg"
-		w = NewWorker(context.WithValue(ctx, "wg", wg), q)
+		w = NewWorker(context.WithValue(ctx, contextKey("wg"), wg), q)
 		go w.Start(service)
 	}
 

--- a/set.go
+++ b/set.go
@@ -101,21 +101,13 @@ func (r *QueryResult) SetMetrics(recs records, valueOnError string) error {
 		return errors.New("sub-metrics are not compatible with data-field")
 	}
 
-	submetrics := map[string]string{}
-
-	if len(r.Query.SubMetrics) > 0 {
-		submetrics = r.Query.SubMetrics
-	} else {
-		submetrics = map[string]string{"": r.Query.DataField}
-	}
-
 	// We need to make sure not to default to a value on error before
 	// it has been registered once before since re-registering might
 	// not work if different labels are used.
-	metricSet := false
-	for k := range r.Result {
-		if strings.HasPrefix(k, r.generateMetricName("")) {
-			if len(recs) == 0 && valueOnError != "" {
+	if len(recs) == 0 && valueOnError != "" {
+		metricSet := false
+		for k := range r.Result {
+			if strings.HasPrefix(k, r.generateMetricName("")) {
 				err := setValueForResult(r.Result[k], valueOnError)
 				if err != nil {
 					return err
@@ -123,9 +115,17 @@ func (r *QueryResult) SetMetrics(recs records, valueOnError string) error {
 				metricSet = true
 			}
 		}
+		if metricSet {
+			return nil
+		}
 	}
-	if metricSet {
-		return nil
+
+	submetrics := map[string]string{}
+
+	if len(r.Query.SubMetrics) > 0 {
+		submetrics = r.Query.SubMetrics
+	} else {
+		submetrics = map[string]string{"": r.Query.DataField}
 	}
 
 	facetsWithResult := make(map[string]metricStatus, 0)

--- a/set.go
+++ b/set.go
@@ -10,11 +10,13 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
-// MetricStatus indicate if metric is registered or unregistered
-type MetricStatus int
+type record map[string]interface{}
+type records []record
+
+type metricStatus int
 
 const (
-	registered MetricStatus = iota
+	registered metricStatus = iota
 	unregistered
 )
 
@@ -34,21 +36,30 @@ func NewQueryResult(q *Query) *QueryResult {
 	return r
 }
 
-func (r *QueryResult) registerMetric(facets map[string]interface{}, suffix string) (string, MetricStatus) {
-	labels := prometheus.Labels{}
+func (r *QueryResult) generateMetricName(suffix string) string {
 	metricName := r.Query.Name
 	if suffix != "" {
 		metricName = fmt.Sprintf("%s_%s", r.Query.Name, suffix)
 	}
+	return metricName
+}
 
+func (r *QueryResult) generateMetricUniqueKey(facets map[string]interface{}, suffix string) string {
 	jsonData, _ := json.Marshal(facets)
-	resultKey := fmt.Sprintf("%s%s", metricName, string(jsonData))
+	return fmt.Sprintf("%s%s", r.generateMetricName(suffix), string(jsonData))
+}
 
+func (r *QueryResult) createMetric(facets map[string]interface{}, suffix string) (string, metricStatus) {
+	metricName := r.generateMetricName(suffix)
+	resultKey := r.generateMetricUniqueKey(facets, suffix)
+
+	labels := prometheus.Labels{}
 	for k, v := range facets {
 		labels[k] = strings.ToLower(fmt.Sprintf("%v", v))
 	}
 
-	if _, ok := r.Result[resultKey]; ok { // A metric with this name is already registered
+	if _, ok := r.Result[resultKey]; ok {
+		// A metric with this key is already created and assumed to be registered
 		return resultKey, registered
 	}
 
@@ -60,9 +71,6 @@ func (r *QueryResult) registerMetric(facets map[string]interface{}, suffix strin
 	})
 	return resultKey, unregistered
 }
-
-type record map[string]interface{}
-type records []record
 
 func setValueForResult(r prometheus.Gauge, v interface{}) error {
 	switch t := v.(type) {
@@ -83,16 +91,17 @@ func setValueForResult(r prometheus.Gauge, v interface{}) error {
 }
 
 // SetMetrics set and register metrics
-func (r *QueryResult) SetMetrics(recs records) (map[string]MetricStatus, error) {
+func (r *QueryResult) SetMetrics(recs records, valueOnError string) error {
 	// Queries that return only one record should only have one column
 	if len(recs) > 1 && len(recs[0]) == 1 {
-		return nil, errors.New("There is more than one row in the query result - with a single column")
+		return errors.New("There is more than one row in the query result - with a single column")
 	}
 
 	if r.Query.DataField != "" && len(r.Query.SubMetrics) > 0 {
-		return nil, errors.New("sub-metrics are not compatible with data-field")
+		return errors.New("sub-metrics are not compatible with data-field")
 	}
 
+	processRecs := recs
 	submetrics := map[string]string{}
 
 	if len(r.Query.SubMetrics) > 0 {
@@ -101,8 +110,27 @@ func (r *QueryResult) SetMetrics(recs records) (map[string]MetricStatus, error) 
 		submetrics = map[string]string{"": r.Query.DataField}
 	}
 
-	facetsWithResult := make(map[string]MetricStatus, 0)
-	for _, row := range recs {
+	// We need to make sure not to default to a value on error before
+	// it has been registered once before since re-registering might
+	// not work if different labels are used.
+	metricSet := false
+	for k := range r.Result {
+		if strings.HasPrefix(k, r.generateMetricName("")) {
+			if len(recs) == 0 && valueOnError != "" {
+				err := setValueForResult(r.Result[k], valueOnError)
+				if err != nil {
+					return err
+				}
+				metricSet = true
+			}
+		}
+	}
+	if metricSet {
+		return nil
+	}
+
+	facetsWithResult := make(map[string]metricStatus, 0)
+	for _, row := range processRecs {
 		for suffix, datafield := range submetrics {
 			facet := make(map[string]interface{})
 			var (
@@ -123,7 +151,7 @@ func (r *QueryResult) SetMetrics(recs records) (map[string]MetricStatus, error) 
 					}
 				} else { // this is the actual gauge data
 					if dataFound {
-						return nil, errors.New("Data field not specified for multi-column query")
+						return errors.New("Data field not specified for multi-column query")
 					}
 					dataVal = v
 					dataFound = true
@@ -131,23 +159,23 @@ func (r *QueryResult) SetMetrics(recs records) (map[string]MetricStatus, error) 
 			}
 
 			if !dataFound {
-				return nil, errors.New("Data field not found in result set")
+				return errors.New("Data field not found in result set")
 			}
 
-			key, status := r.registerMetric(facet, suffix)
+			key, status := r.createMetric(facet, suffix)
 			err := setValueForResult(r.Result[key], dataVal)
 			if err != nil {
-				return nil, err
+				return err
 			}
 			facetsWithResult[key] = status
 		}
 	}
-
-	return facetsWithResult, nil
+	r.registerMetrics(facetsWithResult)
+	return nil
 }
 
 // RegisterMetrics registers and unregister gauges
-func (r *QueryResult) RegisterMetrics(facetsWithResult map[string]MetricStatus) {
+func (r *QueryResult) registerMetrics(facetsWithResult map[string]metricStatus) {
 	for key, m := range r.Result {
 		status, ok := facetsWithResult[key]
 		if !ok {

--- a/set.go
+++ b/set.go
@@ -101,7 +101,6 @@ func (r *QueryResult) SetMetrics(recs records, valueOnError string) error {
 		return errors.New("sub-metrics are not compatible with data-field")
 	}
 
-	processRecs := recs
 	submetrics := map[string]string{}
 
 	if len(r.Query.SubMetrics) > 0 {
@@ -130,7 +129,7 @@ func (r *QueryResult) SetMetrics(recs records, valueOnError string) error {
 	}
 
 	facetsWithResult := make(map[string]metricStatus, 0)
-	for _, row := range processRecs {
+	for _, row := range recs {
 		for suffix, datafield := range submetrics {
 			facet := make(map[string]interface{})
 			var (

--- a/set.go
+++ b/set.go
@@ -10,19 +10,21 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
-type metricStatus int
+// MetricStatus indicate if metric is registered or unregistered
+type MetricStatus int
 
 const (
-	registered metricStatus = iota
+	registered MetricStatus = iota
 	unregistered
 )
 
+// QueryResult contains query results
 type QueryResult struct {
 	Query  *Query
 	Result map[string]prometheus.Gauge // Internally we represent each facet with a JSON-encoded string for simplicity
 }
 
-// NewSetMetrics initializes a new metrics collector.
+// NewQueryResult initializes a new metrics collector.
 func NewQueryResult(q *Query) *QueryResult {
 	r := &QueryResult{
 		Query:  q,
@@ -32,7 +34,7 @@ func NewQueryResult(q *Query) *QueryResult {
 	return r
 }
 
-func (r *QueryResult) registerMetric(facets map[string]interface{}, suffix string) (string, metricStatus) {
+func (r *QueryResult) registerMetric(facets map[string]interface{}, suffix string) (string, MetricStatus) {
 	labels := prometheus.Labels{}
 	metricName := r.Query.Name
 	if suffix != "" {
@@ -80,7 +82,8 @@ func setValueForResult(r prometheus.Gauge, v interface{}) error {
 	return nil
 }
 
-func (r *QueryResult) SetMetrics(recs records) (map[string]metricStatus, error) {
+// SetMetrics set and register metrics
+func (r *QueryResult) SetMetrics(recs records) (map[string]MetricStatus, error) {
 	// Queries that return only one record should only have one column
 	if len(recs) > 1 && len(recs[0]) == 1 {
 		return nil, errors.New("There is more than one row in the query result - with a single column")
@@ -98,7 +101,7 @@ func (r *QueryResult) SetMetrics(recs records) (map[string]metricStatus, error) 
 		submetrics = map[string]string{"": r.Query.DataField}
 	}
 
-	facetsWithResult := make(map[string]metricStatus, 0)
+	facetsWithResult := make(map[string]MetricStatus, 0)
 	for _, row := range recs {
 		for suffix, datafield := range submetrics {
 			facet := make(map[string]interface{})
@@ -143,7 +146,8 @@ func (r *QueryResult) SetMetrics(recs records) (map[string]metricStatus, error) 
 	return facetsWithResult, nil
 }
 
-func (r *QueryResult) RegisterMetrics(facetsWithResult map[string]metricStatus) {
+// RegisterMetrics registers and unregister gauges
+func (r *QueryResult) RegisterMetrics(facetsWithResult map[string]MetricStatus) {
 	for key, m := range r.Result {
 		status, ok := facetsWithResult[key]
 		if !ok {

--- a/set_test.go
+++ b/set_test.go
@@ -116,7 +116,7 @@ gauge: <
 }
 
 func (opts *testQuerySetOptions) testQuerySet(t *testing.T) {
-	_, err := opts.q.SetMetrics(opts.rec)
+	err := opts.q.SetMetrics(opts.rec, "")
 	if err != nil {
 		t.Errorf("Error while setting metrics: %v", err)
 		return

--- a/worker.go
+++ b/worker.go
@@ -26,6 +26,7 @@ var defaultBackoff = backoff.Backoff{
 	Factor: 2,
 }
 
+// Worker is responsible for fetching data via SQL Agent
 type Worker struct {
 	query   *Query
 	payload []byte
@@ -36,7 +37,7 @@ type Worker struct {
 	ctx     context.Context
 }
 
-func (w *Worker) SetMetrics(recs records) {
+func (w *Worker) setMetrics(recs records) {
 	list, err := w.result.SetMetrics(recs)
 	if err != nil {
 		w.log.Printf("Error setting metrics: %s", err)
@@ -46,7 +47,8 @@ func (w *Worker) SetMetrics(recs records) {
 	w.result.RegisterMetrics(list)
 }
 
-func (w *Worker) Fetch(url string) (records, error) {
+// Fetch connect to SQL Agent and fetches records
+func (w *Worker) fetch(url string) (records, error) {
 	var (
 		t    time.Time
 		err  error
@@ -83,7 +85,7 @@ func (w *Worker) Fetch(url string) (records, error) {
 		}
 
 		if w.query.ValueOnError != "" {
-			w.SetMetrics([]record{
+			w.setMetrics([]record{
 				map[string]interface{}{
 					"error": w.query.ValueOnError,
 				},
@@ -114,14 +116,15 @@ func (w *Worker) Fetch(url string) (records, error) {
 		return nil, err
 	}
 
-	w.SetMetrics(recs)
+	w.setMetrics(recs)
 
 	return recs, nil
 }
 
+// Start fetching data from specified URL
 func (w *Worker) Start(url string) {
 	tick := func() {
-		_, err := w.Fetch(url)
+		_, err := w.fetch(url)
 		if err != nil {
 			w.log.Printf("Error fetching records: %s", err)
 			return

--- a/worker.go
+++ b/worker.go
@@ -37,18 +37,19 @@ type Worker struct {
 	ctx     context.Context
 }
 
-func (w *Worker) setMetrics(recs records) {
-	list, err := w.result.SetMetrics(recs)
+func (w *Worker) setQueryResultMetrics(recs records) {
+	err := w.result.SetMetrics(recs, w.query.ValueOnError)
 	if err != nil {
 		w.log.Printf("Error setting metrics: %s", err)
 		return
 	}
-
-	w.result.RegisterMetrics(list)
 }
 
-// Fetch connect to SQL Agent and fetches records
-func (w *Worker) fetch(url string) (records, error) {
+func (w *Worker) queryResultError() {
+	w.setQueryResultMetrics(nil)
+}
+
+func (w *Worker) fetchRecords(url string) error {
 	var (
 		t    time.Time
 		err  error
@@ -83,24 +84,18 @@ func (w *Worker) fetch(url string) (records, error) {
 		if err == nil {
 			break
 		}
+		w.log.Print(err)
 
-		if w.query.ValueOnError != "" {
-			w.setMetrics([]record{
-				map[string]interface{}{
-					"error": w.query.ValueOnError,
-				},
-			})
-		}
+		w.queryResultError()
 
 		// Backoff on an error.
-		w.log.Print(err)
 		d := w.backoff.Duration()
 		w.log.Printf("Backing off for %s", d)
 		select {
 		case <-time.After(d):
 			continue
 		case <-w.ctx.Done():
-			return nil, errors.New("Execution was canceled")
+			return errors.New("Execution was canceled")
 		}
 	}
 
@@ -113,18 +108,18 @@ func (w *Worker) fetch(url string) (records, error) {
 	defer resp.Body.Close()
 
 	if err = json.NewDecoder(resp.Body).Decode(&recs); err != nil {
-		return nil, err
+		return err
 	}
 
-	w.setMetrics(recs)
+	w.setQueryResultMetrics(recs)
 
-	return recs, nil
+	return nil
 }
 
 // Start fetching data from specified URL
 func (w *Worker) Start(url string) {
 	tick := func() {
-		_, err := w.fetch(url)
+		err := w.fetchRecords(url)
 		if err != nil {
 			w.log.Printf("Error fetching records: %s", err)
 			return

